### PR TITLE
Refactor interpolation (and string and regex) handling in lexer

### LINKED ISSUE #3621: Multiline RegExp with string interpolation do not get marked as such.
```coffeescript
a = 'hello'

b = RegExp(".*#{a}[0-9]")
```
Outputs:
```
[ [ 'IDENTIFIER',\n  'a',\n  { first_line: 0, first_column: 0, last_line: 0, last_column: 0 },\n  spaced: true ],
  [ '=',\n  '=',\n  { first_line: 0, first_column: 2, last_line: 0, last_column: 2 },\n  spaced: true ],
  [ 'STRING',\n  '\\'hello\\',\n  { first_line: 0, first_column: 4, last_line: 0, last_column: 10 },\n  newLine: true ],
  [ 'TERMINATOR',\n  '\\n',\n  { first_line: 0, first_column: 11, last_line: 0, last_column: 11 } ],
  [ 'IDENTIFIER',\n  'b',\n  { first_line: 2, first_column: 0, last_line: 2, last_column: 0 },\n  spaced: true ],
  [ '=',\n  '=',\n  { first_line: 2, first_column: 2, last_line: 2, last_column: 2 },\n  spaced: true ],
  [ 'IDENTIFIER',\n  'RegExp',\n  { first_line: 2, first_column: 4, last_line: 2, last_column: 9 } ],
  [ 'CALL_START',\n  '(',\n  { first_line: 2, first_column: 10, last_line: 2, last_column: 10 } ],
  [ '(',\n  '(',\n  { first_line: 2, first_column: 11, last_line: 2, last_column: 11 },\n  origin: [ ',\n    'string interpolation',\n    { first_line: 2, first_column: 14, last_line: 2, last_column: 15 } ] ],
  [ 'STRING',\n  '".*"',\n  { first_line: 2, first_column: 12, last_line: 2, last_column: 13 } ],
  [ '+',\n  '+',\n  { first_line: 2, first_column: 16, last_line: 2, last_column: 16 } ],
  [ 'IDENTIFIER',\n  'a',\n  { first_line: 2, first_column: 16, last_line: 2, last_column: 16 } ],
  [ '+',\n  '+',\n  { first_line: 2, first_column: 18, last_line: 2, last_column: 18 } ],
  [ 'STRING',\n  '"[0-9]"',\n  { first_line: 2, first_column: 18, last_line: 2, last_column: 22 } ],
  [ ')',\n  ')',\n  { first_line: 2, first_column: 24, last_line: 2, last_column: 24 },\n  stringEnd: true ],
  [ 'CALL_END',\n  ')',\n  { first_line: 2, first_column: 24, last_line: 2, last_column: 24 } ],
  [ 'TERMINATOR',\n  '\\n',\n  { first_line: 2, first_column: 25, last_line: 2, last_column: 25 } ] ]
```
But,
```coffeescript
a = 'hello'

c = ///.*#{a}[0-9]///
```
Outputs

```
[ [ 'IDENTIFIER',\n  'a',\n  { first_line: 0, first_column: 0, last_line: 0, last_column: 0 },\n  spaced: true ],
  [ '=',\n  '=',\n  { first_line: 0, first_column: 2, last_line: 0, last_column: 2 },\n  spaced: true ],
  [ 'STRING',\n  '\\'hello\\',\n  { first_line: 0, first_column: 4, last_line: 0, last_column: 10 },\n  newLine: true ],
  [ 'TERMINATOR',\n  '\\n',\n  { first_line: 0, first_column: 11, last_line: 0, last_column: 11 } ],
  [ 'IDENTIFIER',\n  'c',\n  { first_line: 2, first_column: 0, last_line: 2, last_column: 0 },\n  spaced: true ],
  [ '=',\n  '=',\n  { first_line: 2, first_column: 2, last_line: 2, last_column: 2 },\n  spaced: true ],
  [ 'IDENTIFIER',\n  'RegExp',\n  { first_line: 2, first_column: 4, last_line: 2, last_column: 4 } ],
  [ 'CALL_START',\n  '(',\n  { first_line: 2, first_column: 4, last_line: 2, last_column: 4 } ],
  [ 'STRING',\n  '".*"',\n  { first_line: 2, first_column: 7, last_line: 2, last_column: 8 } ],
  [ '+',\n  '+',\n  { first_line: 2, first_column: 4, last_line: 2, last_column: 4 } ],
  [ 'IDENTIFIER',\n  'a',\n  { first_line: 2, first_column: 11, last_line: 2, last_column: 11 } ],
  [ '+',\n  '+',\n  { first_line: 2, first_column: 4, last_line: 2, last_column: 4 } ],
  [ 'STRING',\n  '"[0-9]"',\n  { first_line: 2, first_column: 13, last_line: 2, last_column: 17 } ],
  [ 'CALL_END',\n  ')',\n  { first_line: 2, first_column: 20, last_line: 2, last_column: 20 } ],
  [ 'TERMINATOR',\n  '\\n',\n  { first_line: 2, first_column: 21, last_line: 2, last_column: 21 } ] ]
```

In the multi-line regexp there is no '(' token with an `origin` property after the CALL_START token and no corresponding ')' token with a `stringEnd` property set to true before the CALL_END token.

### PR DESCRIPTION
**Pull request by lydell**
See the commit messages for details.

I’m sorry about the size of the first commit, but I don’t think I could have split it up.

Fixes #3621, #3301, #3394, #3348, #2388, #2321.

### CODE REVIEW

**Comment by vendethiel** [comment]
Did not look at the code yet, but thanks for the tests.

File: src/lexer.coffee [hunk 1]
```diff
@@ -594,27 +537,33 @@ exports.Lexer = class Lexer
...
+        when 'TOKENS'
+          # Push all the tokens in the fake 'TOKENS' token. These already have
+          # sane location data.
+          @tokens.push value...
+        when 'NEOSTRING'
+          # Convert 'NEOSTRING' into 'STRING'.
+          converted = fn token[1], i
+          # Optimize out empty strings. We ensure that the tokens stream always
+          # starts with a string token, though, to make sure that the result
+          # really is a string. TODO: Change this according to #3529.
```
Comment (michaelficarra) [inline src/lexer.coffee]: I don't like leaving TODOs around. This isn't a particularly important one either, so let's remove it.

File: src/lexer.coffee [hunk 2]
```diff
@@ -483,106 +453,79 @@ exports.Lexer = class Lexer
...
+
+    unless str[...end.length] is end
+      @error "missing #{end}"
+
+    {tokens, index: offsetInChunk + end.length}
+
+  # Merge the array `tokens` of the fake token types 'TOKENS' and 'NEOSTRING'
+  # (as returned by `matchWithInterpolations`) into the token stream. The value
+  # of 'NEOSTRING's are converted using `fn` and turned into strings using
+  # `quote` first.  The tokens are wrapped in parenthesis if needed, using
```
Comment (michaelficarra) [inline src/lexer.coffee]: > parenthesis

parentheses

File: src/lexer.coffee [hunk 3]
```diff
@@ -483,106 +453,79 @@ exports.Lexer = class Lexer
...
+      index += 1
+
+      # Remove leading and trailing `{` and `}`.
+      nested.shift()
+      nested.pop()
+
+      # Remove leading 'TERMINATOR' (if any).
+      nested.shift() if nested[0]?[0] is 'TERMINATOR'
+
+      if len = nested.length
```
Comment (michaelficarra) [inline src/lexer.coffee]: Avoid assignment within the test. I also like a more explicit test.

```coffee
len = nested.length
if len > 0
  if len > 1
    ...
  ...
```

Comment (lydell) [inline src/lexer.coffee]: That's how I'd write it too. But here I have just moved code from one place to another. Should I change it anyway?

File: src/lexer.coffee [hunk 4]
```diff
@@ -483,106 +453,79 @@ exports.Lexer = class Lexer
...
-      tokens.push @makeToken('NEOSTRING', str[pi...i], strOffset + pi) if pi < i
-      unless errorToken
-        errorToken = @makeToken '', 'string interpolation', offsetInChunk + i + 1, 2
-      inner = expr[1...-1]
-      if inner.length
-        [line, column] = @getLineAndColumnFromChunk(strOffset + i + 2)
-        nested = new Lexer().tokenize inner, line: line, column: column, rewrite: off
-        popped = nested.pop()
-        popped = nested.shift() if nested[0]?[0] is 'TERMINATOR'
-        if len = nested.length
```
Comment (lydell) [inline src/lexer.coffee]: @michaelficarra here is the original code that I moved.

File: src/lexer.coffee (hunk 3 above)
Comment (michaelficarra) [inline src/lexer.coffee]: Yep.

**Comment by michaelficarra** [comment]
LGTM. Very nice job, especially with all the tests. This was one of the most direly needed features. :+1:.

File: src/lexer.coffee (hunk 3 above)
Comment (lydell) [inline src/lexer.coffee]: The whole file is filled with assignments in tests. Wouldn't it be better with a separate "cleanup" PR?

Comment (michaelficarra) [inline src/lexer.coffee]: If it's involved in a PR, we usually ask that it is cleaned up. I'm okay either way if you're willing to send a cleanup PR.

**Comment by lydell** [comment]
I have fixed all mentioned issues now. Moreover, now it also fixes #3529. Ready for another round of review.

**Comment by michaelficarra** [comment]
LGTM.

**Comment by epidemian** [comment]
This looks very good. Thanks for putting such work on some of the hairy corners of the compiler, @lydell!

**Comment by xixixao** [comment]
@lydell :trophy:

**Comment by jashkenas** [comment]
And let's tag 'em as we close 'em ;)

**Comment by michaelficarra** [comment]
I wasn't sure how to tag a PR that fixes a bug.

**Comment by jashkenas** [comment]
It was a big ol' change/rewrite, so I'm going with "change fixed". If it had just been a single bug or two, then "bug fixed".

**Comment by vendethiel** [comment]
I'm not sure we need the "fixed" tag. I mean, a "bug" PR without the "fixed" tag would mean the PR added a bug on purpose?

**Comment by jashkenas** [comment]
We need it. It's nice to be able to search through Issues **and** PR's consistently, and that's harder if we only keep things organized half the time. On occasion, a minor, but unfixable bug might be closed without fixing, and we want to be able to find those as well.

**Pull request merged**